### PR TITLE
Stefanozampini/small additions

### LINKED ISSUE #644: Including petsc.h redefines MPI calls as PETSc macros
I noticed that when MFEM is built with PETSc, MPI calls like `MPI_Isend`, `MPI_Irecv` are being replaced with PETSc macros. These macros come from `petsclog.h` which is included by the chian `petsc.h` -> `petscbag.h` -> `petscsys.h` -> `petsclog.h`. The comments inside the header indicate that these macros are intended for internal PETSc use: https://github.com/petsc/petsc/blob/0a42d0a183c9f292d6e79078ae7ca0ad5320b535/include/petsclog.h#L300-L303.

@jedbrown, @stefanozampini, do you know if this is the intended behavior?

### PR DESCRIPTION
**Pull request by stefanozampini**
This branch provides few additions to the PETSc classes and few other minor fixes I needed for a pde-constrained optimization code I'm currently working on 
With this branch, petsc.h is not included in petsc.hpp 

Edit: Resolves #644 (among other things).

### CODE REVIEW

File: mesh/mesh.cpp [hunk 1]
```diff
@@ -4878,14 +4882,20 @@ int *Mesh::GeneratePartitioning(int nparts, int part_method)
...
          partitioning[i] = 0;
       }
    }
+   else if (nparts == NumOfElements)
+   {
+      for (i = 0; i < NumOfElements; i++)
+      {
+         partitioning[i] = i;
+      }
+   }
```
Comment (tzanio) [inline mesh/mesh.cpp]: This will be in conflict with https://github.com/mfem/mfem/pull/674/files#diff-7dc3e1a317e9f6a450c7cf9f0f1bc8e3

Comment (stefanozampini) [reply mesh/mesh.cpp]: I'll drop my commit

File: mesh/mesh.cpp [hunk 2]
```diff
@@ -2534,7 +2534,11 @@ Mesh::Mesh(const Mesh &mesh, bool copy_nodes)
    }
 
    // Copy the vertices
-   mesh.vertices.Copy(vertices);
+   vertices.SetSize(NumOfVertices);
+   for (int i = 0; i < NumOfVertices; i++)
+   {
+      vertices[i].SetCoords(spaceDim, mesh.vertices[i]());
+   }
```
Comment (v-dobrev) [inline mesh/mesh.cpp]: Why this change?

Comment (stefanozampini) [reply mesh/mesh.cpp]: Fix a GCC 8.1 warning, see fa6fdb7

Comment (v-dobrev) [reply mesh/mesh.cpp]: Oh, I think I fixed this warning by making class `Vertex` trivially copyable. It used to have an empty, explicitly defined destructor which was making it non-trivially copyable as defined by the c++ standard ([TriviallyCopyable](https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable)). The fix is in 265006d59f which in now merged in master.

Comment (stefanozampini) [reply mesh/mesh.cpp]: I have dropped my commit

File: mesh/mesh.cpp (hunk 1 above)
Comment (stefanozampini) [reply mesh/mesh.cpp]: I have dropped my commit

File: mesh/mesh.cpp (hunk 2 above)
Comment (tzanio) [reply mesh/mesh.cpp]: Thanks Stefano
